main.go

```diff
@@ -106,11 +110,53 @@ func main() {
 	http.Handle("/live", handlers.NewLivenessHandler())
 	http.Handle("/metrics", promhttp.HandlerFor(reg, promhttp.HandlerOpts{}))
 
-	log.Info("Starting http handler...")
-	address := fmt.Sprintf(":%s", Options.ListenPort)
+	// Run listen on http and https ports if HTTPSCertFile/HTTPSKeyFile set
+	var httpsServer, httpServer http.Server
+
+	httpsListen := func() {
+		log.Infof("Starting https handler on %s...", httpsServer.Addr)
+		if err := httpsServer.ListenAndServeTLS(Options.HTTPSCertFile, Options.HTTPSKeyFile); err != http.ErrServerClosed {
+			log.Fatalf("HTTPS listener closed: %v", err)
+		}
+	}
+
+	httpListen := func() {
+		log.Infof("Starting http handler on %s...", httpServer.Addr)
+		if err := httpServer.ListenAndServe(); err != http.ErrServerClosed {
+			log.Fatalf("HTTP listener closed: %v", err)
+		}
+	}
+
+	// Interrupt servers on SIGINT/SIGTERM
+	stop := make(chan os.Signal, 1)
+	signal.Notify(stop, os.Interrupt, syscall.SIGINT, syscall.SIGTERM)
+
 	if Options.HTTPSKeyFile != "" && Options.HTTPSCertFile != "" {
-		log.Fatal(http.ListenAndServeTLS(address, Options.HTTPSCertFile, Options.HTTPSKeyFile, nil))
+		httpsServer = http.Server{
+			Addr: fmt.Sprintf(":%s", Options.ListenPort),
+		}
+		go httpsListen()
+		if Options.HTTPListenPort != "" {
+			httpServer = http.Server{
+				Addr: fmt.Sprintf(":%s", Options.HTTPListenPort),
+			}
+			go httpListen()
+		}
+	} else {
+		httpServer = http.Server{
+			Addr: fmt.Sprintf(":%s", Options.ListenPort),
+		}
+		go httpListen()
+	}
+	<-stop
+	if err := httpsServer.Shutdown(context.TODO()); err != nil {
+		log.Fatalf("HTTPS shutdown failed: %v", err)
+	} else {
+		log.Info("HTTPS server terminated gracefully")
+	}
+	if err := httpServer.Shutdown(context.TODO()); err != nil {
```

Comment:
These two calls should check for nil I think

Reply:
`Shutdown` returning non-nil is an error, I think we want to have this logged at least

Reply:
No, I mean `httpServer` could be nil. This will segfault

Reply:
Hmm, it appears `http.Server` cannot be nil and it won't crash on `Shutdown` when not running - https://go.dev/play/p/Hz-hDxEZlOf

Reply:
Perfect. Just went to look it up and saw:
> The zero value for Server is a valid configuration

ref: https://pkg.go.dev/net/http@go1.18.1#Server

:tada: 

>There is no branch where httpServer could be nil 

Not that it matters now, but it would have been unset in the case where we set the https config, but not http port, right?
Even still it's clear to me now that `http.Server` is not a pointer value so would never be nil for that reason.